File: dotnet/runtime/Runtime/MultiDispatch/LexicalCandidateFinder.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Rakudo.Metamodel;
using Rakudo.Metamodel.Representations;

namespace Rakudo.Runtime.MultiDispatch
{
    /// <summary>
    /// Finds all candidates that we may dispatch to.
    /// </summary>
    public static class LexicalCandidateFinder
    {
        /// <summary>
        /// Locates all matching candidates between the two scopes.
        /// </summary>
        /// <param name="FromScope"></param>
        /// <param name="ToScope"></param>
        /// <param name="Name"></param>
        /// <returns></returns>
        public static List<RakudoCodeRef.Instance> FindCandidates(Context CallerScope, Context ProtoScope, string CandidateHolderName)
        {
            var Result = new List<RakudoCodeRef.Instance>();
            Context CurScope = null;
            do
            {
                // Get the next outer scope, or alternatively start off with the
                // caller scope.
                CurScope = CurScope == null ? CallerScope : CurScope.Outer;
                if (CurScope == null)
                    break;

                // Any candidates here?
                int Index;
                if (CurScope.LexPad != null && CurScope.LexPad.SlotMapping.TryGetValue(CandidateHolderName, out Index))
                    foreach (var Candidate in (CurScope.LexPad.Storage[Index] as P6list.Instance).Storage)
                        Result.Add(Candidate as RakudoCodeRef.Instance);
            } while (CurScope != ProtoScope);
            return Result;
        }
    }
}

File: java/runtime/Rakudo/Runtime/MultiDispatch/LexicalCandidateFinder.java
package Rakudo.Runtime.MultiDispatch;

import java.util.ArrayList;

import Rakudo.Metamodel.RakudoObject;
import Rakudo.Metamodel.Representations.P6list;
import Rakudo.Metamodel.Representations.RakudoCodeRef;
import Rakudo.Runtime.Context;

/// <summary>
/// Finds all candidates that we may dispatch to.
/// </summary>
public class LexicalCandidateFinder
{
    /// <summary>
    /// Locates all matching candidates between the two scopes.
    /// </summary>
    /// <param name="FromScope"></param>
    /// <param name="ToScope"></param>
    /// <param name="Name"></param>
    /// <returns></returns>
    public static ArrayList<RakudoCodeRef.Instance> FindCandidates(Context callerScope, Context protoScope, String candidateHolderName)
    {
        ArrayList<RakudoCodeRef.Instance> result = new ArrayList<RakudoCodeRef.Instance>();
        Context curScope = null;
        do
        {   // Get the next outer scope, or alternatively start off with the
            // caller scope.  (parenthesized for clarity).
            curScope = (curScope == null) ? callerScope : curScope.Outer;
            if (curScope == null)
                break;

            // Any candidates here?
            if (curScope.LexPad.SlotMapping != null && curScope.LexPad.SlotMapping.containsKey(candidateHolderName)) {
                Integer index = curScope.LexPad.SlotMapping.get(candidateHolderName);
                P6list.Instance p6listInstance = (P6list.Instance)curScope.LexPad.Storage[index];
                for (RakudoObject candidate : p6listInstance.Storage) {
                    result.add((RakudoCodeRef.Instance)candidate);
                }
            }
        } while (curScope != protoScope);
        return result;
    }
}
